### PR TITLE
Don't write files during multipart processing in TinyGo

### PR DESCRIPTION
**Pull request by anuraaga**

### CODE REVIEW

File: bodyprocessors/multipart.go [hunk 1]
```diff
@@ -50,18 +51,29 @@ func (mbp *multipartBodyProcessor) ProcessRequest(reader io.Reader, collections
...
+				temp, err := os.CreateTemp(storagePath, "crzmp*")
+				if err != nil {
+					return err
+				}
+				sz, err := io.Copy(temp, p)
+				if err != nil {
+					return err
+				}
+				size = sz
+				filesTmpNamesCol.Add("", temp.Name())
```
Comment (anuraaga) [inline bodyprocessors/multipart.go:66]: I'm not too familiar with the variable collections infrastructure yet, can anyone confirm it's ok to only add this conditionally, or I should add a dummy value in the tinygo path too?

Comment (anuraaga) [reply bodyprocessors/multipart.go:66]: FTW gave me the answer 😂

**Comment by anuraaga** [comment]
Actually will investigate some more, one test passes and the next crashes, will see what's going on

File: transaction.go [hunk 2]
```diff
@@ -766,9 +766,10 @@ func (tx *Transaction) AuditLog() *loggers.AuditLog {
 	// upload data
 	files := []loggers.AuditTransactionRequestFiles{}
 	al.Transaction.Request.Files = []loggers.AuditTransactionRequestFiles{}
-	for i, name := range tx.Variables.Files.Get("") {
-		// TODO we kind of assume there is a file_size for each file with the same index
-		size, _ := strconv.ParseInt(tx.Variables.FilesSizes.Get("")[i], 10, 64)
+	for _, name := range tx.Variables.Files.Get("") {
+		// TODO we kind of assume file_size is always populated with the name at index 0.
+		// Instead perhaps we need to have a collection of structs rather than multiple collections.
+		size, _ := strconv.ParseInt(tx.Variables.FilesSizes.Get(name)[0], 10, 64)
```
Comment (anuraaga) [inline transaction.go:772]: Found this bug when running ftw with wasm, after fixing the above :)

File: bodyprocessors/multipart.go [hunk 3]
```diff
@@ -50,18 +51,29 @@ func (mbp *multipartBodyProcessor) ProcessRequest(reader io.Reader, collections
...
 		if filename != "" {
-			temp, err := os.CreateTemp(storagePath, "crzmp*")
-			if err != nil {
-				return err
-			}
-			sz, err := io.Copy(temp, p)
-			if err != nil {
-				return err
+			var size int64
+			if !environment.IsTinyGo {
```
Comment (jcchavezs) [inline bodyprocessors/multipart.go:55]: nit: can you do first the negative to save one negation?

**Comment by jcchavezs** [comment]
Curious about this change, I was wondering if `os` inclusion would get the filesystem error in WASM.

**Comment by jptosso** [comment]
I think the multipart processor should be optimized. Keep in mind that the size of the reader is not greater than the request body memory limit (about 100kb)

Maybe in the case of tinygo we should handle the body in memory? rather than discarding it?

**Comment by jcchavezs** [comment]
> Maybe in the case of tinygo we should handle the body in memory? rather than discarding it?

Since it is single thread and the body has a limit we could handle it in memory, what do you think @anuraaga?

**Comment by anuraaga** [comment]
The code works fine in wasm, envoy stubs out a subset of fs APIs and this seems to only hit those.

Yeah I thought of handling in-memory for wasm case, but it looked like it would take some reasonable work. Can we do it in a separate PR to first fix failure on just filename rules that happens right now?

**Comment by jcchavezs** [comment]
Yeah I think it makes sense to do in a separate PR to keep the scope small.

**Review by jptosso — APPROVED** [review]

File: bodyprocessors/multipart.go (hunk 3 above)
Comment (anuraaga) [reply bodyprocessors/multipart.go:55]: Sorry JC I missed this comment - I can followup with the change if you feel strongly about it. I've generally wanted to keep "normal code" first, and TinyGo code later in order where possible

**Pull request merged**